torchtitan/datasets/hf_datasets.py

```diff
@@ -99,32 +101,61 @@ def __init__(
         self.seq_len = seq_len
         self.infinite = infinite
 
+        # variables for checkpointing
+        self._sample_idx = 0
+        self._all_tokens: List[int] = []
+        self._first_iter = True
+
     def __iter__(self):
         max_buffer_token_len = 1 + self.seq_len
-        all_tokens: List[int] = []
 
         while True:
-            for sample in iter(self._data):
+            # Skip samples for the first iteration to resume from checkpoint
+            data_iter = (
+                self._get_skipped_data_iter(self._sample_idx)
+                if self._first_iter
+                else iter(self._data)
+            )
+            self._first_iter = False
```

Comment:
Let's put all of them into the helper function to make it cleaner, e.g. with name `_get_data_iter` or `_get_maybe_skipped_data_iter`. Also this function doesn't need input, because `self._sample_idx` and `self._first_iter` are both class variables.